[package] ignore dunder import errors

lobstergrindset opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#61148**
* #61147
* #61146

Changes `__import__` processing to silently skip cases where the `__import__` statement cannot be parsed

Differential Revision: [D29559680](https://our.internmc.facebook.com/intern/diff/D29559680)

Commit:
[package] ignore dunder import errors

[ghstack-poisoned]

Commit:
Update on "[package] ignore dunder import errors"


Changes `__import__` processing to silently skip cases where the `__import__` statement cannot be parsed. Adds failed imports to a list retrievable by `PackageExporter.failed_dunder_import_list()`.

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29559680).

suo commented on torch/package/package_exporter.py:
```diff
@@ -970,6 +974,15 @@ def relied_on_by(self, module_name: str) -> List[str]:
         else:
             return []
 
+    def failed_dunder_import_list(self) -> List[str]:
```
I'm not sure what discoverability for this API will be. Ideally we'd like the API to be "obvious"—someone will look in the API reference pretty and quickly find what they're looking for. The methods introduced in the PR below are good examples of obvious APIs—they correspond to actions that are common in the rest the API and it's pretty clear what they do even from the name.

But `failed_dunder_import_list` is pretty inscrutable—I'm not sure how a user would figure out what this method does, or how they would know to call it, or what they would do with results. Maybe we should just eat the `__import__` error and call it a day?

Commit:
Update on "[package] ignore dunder import errors"


Changes `__import__` processing to silently skip cases where the `__import__` statement cannot be parsed. Adds failed imports to a list retrievable by `PackageExporter.failed_dunder_import_list()`.

Differential Revision: [D29559680](https://our.internmc.facebook.com/intern/diff/D29559680)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29559680).

suo commented on torch/package/find_file_dependencies.py:
```diff
@@ -57,47 +57,46 @@ def _grab_node_str(self, node):
         if hasattr(node.func, "id") and node.func.id == "__import__":
-            if type(node.args[0]) not in [ast.Constant, ast.Str]:
-                # We don't want to parse dynamic uses of __import__
+            try:
```
can we narrow the check? we really only want to catch when the imported module isn't constant, any other exceptions should get raised properly. It seems like the existing check only checks the first argument for constantness—maybe the narrowest check would be to check all arguments for constantness and skip if any of them aren't.

Lilyjjo replied:
That might be too strict of a check -- The global/local variables aren't constants normally ex: `module = __import__(module_name, globals(), locals(), [''])`

suo replied:
alright, it's probably not so bad anyway since the try-catch only covers `__import__` calls
